src/hooks/useMaxSiblingWidth.ts

```diff
@@ -0,0 +1,68 @@
+import { useCallback, useEffect, useLayoutEffect, useState } from 'react'
+import { useSelector } from 'react-redux'
+import Path from '../@types/Path'
+import { findAnyChild } from '../selectors/getChildren'
+import editingValueStore from '../stores/editingValue'
+import hashPath from '../util/hashPath'
+import head from '../util/head'
+import parentOf from '../util/parentOf'
+
+/**
+ * Helper function to collect nodes with the specified `data-parent-path`.
+ * If no thought node is found, it continues searching with `data-grandparent-path`.
+ */
+const collectNodes = (path: Path, hasNonDividerSiblings: boolean): HTMLElement[] => {
+  let collectedNodes: HTMLElement[] = []
+  /** Collect nodes by provided attribute and value. */
+  const collectByAttribute = (attribute: string, value: string): HTMLElement[] =>
+    Array.from(document.querySelectorAll<HTMLElement>(`.tree-node[data-${attribute}='${value}']:not(.divider-thought)`))
+
+  if (hasNonDividerSiblings) {
+    collectedNodes = collectByAttribute('parent-path', hashPath(path))
+  } else {
+    const grandparentPathArray = parentOf(path)
+    // Ensure grandParentPathArray is a non-empty array
+    if (Array.isArray(grandparentPathArray) && grandparentPathArray.length) {
+      const grandparentPath = hashPath(grandparentPathArray)
+      collectedNodes = collectByAttribute('grandparent-path', grandparentPath)
+    }
+  }
+
+  return collectedNodes
+}
+
+/**
+ * Helper function to calculate maximum width among '.thought' elements.
+ */
+const calculateMaxWidth = (elements: HTMLElement[]): number => {
+  return elements.reduce((maxWidth, el) => {
+    const thought = el.querySelector<HTMLElement>('.thought')
+    const width = thought ? thought.offsetWidth : 0
+    return Math.max(maxWidth, width)
+  }, 0)
+}
+
+/**
+ * Hook to calculate the width of a divider, equal to the max width of a sibling
+ * at a particular `data-parent-path`.
+ */
+const useMaxSiblingWidth = (parentPath: Path): number => {
+  const [maxWidth, setMaxWidth] = useState(0)
+  const hasNonDividerSiblings = useSelector(state =>
+    findAnyChild(state, head(parentPath), child => child.value !== '---'),
+  )
+
+  const updateMaxWidth = useCallback(() => {
+    const nodes = collectNodes(parentPath, !!hasNonDividerSiblings)
+    setMaxWidth(calculateMaxWidth(nodes))
+  }, [parentPath, hasNonDividerSiblings])
+
+  const editingVal = editingValueStore.useSelector()
```

Comment:
This will cause the render function to be called on every keystroke, which is probably more than necessary. `updateMaxWidth` can be called outside of the component lifecycle and the component will only re-render if it ends up changing `maxWidth`. Hence the reason for `editingValueStore.useEffect`. Maybe we need to add `editingValueStore.useLayoutEffect`.

Reply:
Tried adding this to react-ministore.ts

`const useChangeLayoutEffect = (cb: (state: T) => void) => useLayoutEffect(() => store.subscribe(cb), [cb])`

This does not work for the paste and delete cases

Reply:
Okay. Let me know if you have any other ideas of how to do this without the render function getting called on every keystroke.

Reply:
`editingValueStore.useEffect(updateMaxWidth)`

@raineorshine For the line above, I think the callback gets called before render. So the calculations don't work until the next cycle. Same for useLayoutEffect.

Some how need to run the logic after the render, because the logic is dependent on calculations from the actual DOM and not the virtual DOM

Reply:
> @raineorshine For the line above, I think the callback gets called before render. So the calculations don't work until the next cycle. Same for useLayoutEffect.

Yes, that's what I have observed as well unfortunately.